### PR TITLE
Add SIAMFANLEquations wrapper

### LINKED ISSUE #109: Wrap SIAMFANLEquations
https://github.com/ctkelley/SIAMFANLEquations.jl

### PR DESCRIPTION
**Pull request by ErikQQY**
Fix #109

## Checklist

- [x] Newton and PT solvers
- [x] JFNK linear solvers
- [x] Documentations
- [x] Test cases
- [x] Optimize NLstats, jacobian, and others

### CODE REVIEW

File: ext/NonlinearSolveSIAMFANLEquationsExt.jl [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        if res.errcode == 0
+            retcode = ReturnCode.Success
+        elseif res.errcode == 10
+            retcode = ReturnCode.MaxIters
+        elseif res.errcode == 1
+            retcode = ReturnCode.Failure
+            @error("Line search failed")
+        elseif res.errcode == -1
+            retcode = ReturnCode.Default
+            @info("Initial iterate satisfies the termination criteria")
```
Comment (ChrisRackauckas) [inline ext/NonlinearSolveSIAMFANLEquationsExt.jl:42]:
```suggestion
        elseif res.errcode == -1
            retcode = ReturnCode.Success
```

Comment (ChrisRackauckas) [reply ext/NonlinearSolveSIAMFANLEquationsExt.jl:42]: No error throwing, just return codes

File: ext/NonlinearSolveSIAMFANLEquationsExt.jl [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    N = length(u)
+    FS = zeros(eltype(u), N)
+    FPS = zeros(eltype(u), N, N)
+    # Allocate ahead for Krylov basis
+
+    # Jacobian free Newton Krylov
+    if linsolve !== nothing
+        JVS = linsolve == :gmres ? zeros(eltype(u), N, 3) : zeros(eltype(u), N)
+        # `linsolve` as a Symbol to keep unified interface with other EXTs, SIAMFANLEquations directly use String to choose between linear solvers
+        linsolve_alg = strip(repr(linsolve), ':')
```
Comment (avik-pal) [inline ext/NonlinearSolveSIAMFANLEquationsExt.jl:72]: this can be just `String(linsolve)` right?

File: src/extension_algs.jl [hunk 3]
```diff
@@ -206,3 +206,35 @@ function NLsolveJL(; method = :trust_region, autodiff = :central, store_trace =
...
+  - `autodiff`: the choice of method for generating the Jacobian. Defaults to `:central` or
+    central differencing via FiniteDiff.jl. The other choices are `:forward`.
+  - `show_trace`: whether to show the trace.
+  - `delta`: initial pseudo time step, default is 1e-3.
+  - `linsolve` : JFNK linear solvers, choices are `gmres` and `bicgstab`.
+
+### Submethod Choice
+
+  - `:newton`: Classical Newton method.
+  - `:pseudotransient`:
```
Comment (avik-pal) [inline src/extension_algs.jl:225]: missing docs

File: src/extension_algs.jl [hunk 4]
```diff
@@ -206,3 +206,35 @@ function NLsolveJL(; method = :trust_region, autodiff = :central, store_trace =
     return NLsolveJL(method, autodiff, store_trace, extended_trace, linesearch, linsolve,
         factor, autoscale, m, beta, show_trace)
 end
+
+"""
+    SIAMFANLEquationsJL(; method = :newton, autodiff = :central)
```
Comment (avik-pal) [inline src/extension_algs.jl:211]: needs the other kwargs

File: ext/NonlinearSolveSIAMFANLEquationsExt.jl [hunk 5]
```diff
@@ -0,0 +1,171 @@
+module NonlinearSolveSIAMFANLEquationsExt
+
+using NonlinearSolve, SciMLBase
+using SIAMFANLEquations
+import ConcreteStructs: @concrete
+import UnPack: @unpack
+import FiniteDiff, ForwardDiff
+
+function SciMLBase.__solve(prob::NonlinearProblem, alg::SIAMFANLEquationsJL, args...; abstol = 1e-8,
+        reltol = 1e-8, alias_u0::Bool = false, maxiters = 1000, termination_condition = nothing, kwargs...)
```
Comment (ChrisRackauckas) [inline ext/NonlinearSolveSIAMFANLEquationsExt.jl:10]: Is this the same default we are using elsewhere?

File: ext/NonlinearSolveSIAMFANLEquationsExt.jl [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+                fu_cache = similar(fu)
+                function (J, x, p)
+                    uf.p = p
+                    ForwardDiff.jacobian!(J, uf, fu_cache, x, cache)
+                    return J
+                end
+            else
+                function (J, x, p)
+                    uf.p = p
+                    FiniteDiff.finite_difference_jacobian!(J, uf, x, cache)
```
Comment (ChrisRackauckas) [inline ext/NonlinearSolveSIAMFANLEquationsExt.jl:119]: Why is this not just reusing the Jacobian machinery in this package?

**Comment by ChrisRackauckas** [comment]
Just two comments to address.

File: ext/NonlinearSolveSIAMFANLEquationsExt.jl (hunk 5 above)
Comment (avik-pal) [reply ext/NonlinearSolveSIAMFANLEquationsExt.jl:10]: The `abstol` and `reltol` are determined based on the problem eltype

Comment (ErikQQY) [reply ext/NonlinearSolveSIAMFANLEquationsExt.jl]: Yeah, ```abstol``` and ```reltol``` should just be determined from the problem eltype

File: ext/NonlinearSolveSIAMFANLEquationsExt.jl (hunk 6 above)
Comment (ErikQQY) [reply ext/NonlinearSolveSIAMFANLEquationsExt.jl]: Just refactored to use the built-in Jacobian machinery

File: ext/NonlinearSolveSIAMFANLEquationsExt.jl [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        retcode = ReturnCode.MaxIters
+    elseif sol.errcode == 1
+        retcode = ReturnCode.Failure
+    elseif sol.errcode == -1
+        retcode = ReturnCode.Default
+    end
+
+    # pseudo transient continuation has a fixed cost per iteration, iteration statistics are not interesting here.
+    stats = method == :pseudotransient ? nothing : (SciMLBase.NLStats(sum(sol.stats.ifun), sum(sol.stats.ijac), 0, 0, sum(sol.stats.iarm)))
+    println(sol.stats)
```
Comment (avik-pal) [inline ext/NonlinearSolveSIAMFANLEquationsExt.jl:137]: needs to be deletec?

Comment (ErikQQY) [reply ext/NonlinearSolveSIAMFANLEquationsExt.jl]: oops, just deleted

**Review by ChrisRackauckas — APPROVED** [review]
I think this is good. Avik can follow up if he sees anything else.

**Pull request merged**